client/client.go

```diff
@@ -131,14 +132,34 @@ func (a *ApiClient) Request(code AuthToken) *ApiRequest {
 type ApiRequest struct {
 	api *ApiClient
 	// authorization code to use for requests
-	auth AuthToken
+	auth   AuthToken
+	revoke func() (AuthToken, error)
 }
 
 func (ar *ApiRequest) Do(req *http.Request) (*http.Response, error) {
 	if req.Header.Get("Authorization") == "" {
 		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", ar.auth))
 	}
-	return ar.api.Do(req)
+	r, err := ar.api.Do(req)
+	if r != nil && r.StatusCode == http.StatusUnauthorized {
+		// invalid JWT-token
+		log.Info("Device unauthorized; attempting reauthorization")
+		if jwt, e := ar.revoke(); e == nil {
+			ar.auth = jwt
+			if req.GetBody != nil {
```

Comment:
Won't this be set always since we are doing client requests only? Shouldn't we check for `req.Body` here and copy only if we have some body?

Reply:
I think that's what tried first, but it turns out that the http.Do() call "wipes out" the body, and the only way to get it back is through the (optional) GetBody function in the Request structure.